File: backend/app/schemas/utility_billing.py
"""
Utility Billing, Tariffs & Cost Allocation — Pydantic v2 Schemas
"""
from __future__ import annotations

import uuid
from datetime import date
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, field_validator, model_validator


# ── Tariff schemas ─────────────────────────────────────────────────────────────

class TariffCreate(BaseModel):
    tariff_code:   str
    name:          str
    description:   Optional[str] = None
    utility_type:  str
    tariff_type:   str = "FLAT"
    supplier_id:   Optional[uuid.UUID] = None
    effective_from: date
    effective_to:  Optional[date] = None
    unit:          str
    currency_code: str = "USD"
    base_rate:     Decimal
    peak_rate:     Optional[Decimal] = None
    offpeak_rate:  Optional[Decimal] = None
    demand_rate:   Optional[Decimal] = None
    fixed_charge:  Optional[Decimal] = None
    tax_rate_pct:  Optional[Decimal] = None
    tax_rule:      Optional[str] = None
    is_active:     bool = True
    notes:         Optional[str] = None

    @field_validator("utility_type", "tariff_type", mode="before")
    @classmethod
    def upper(cls, v: str) -> str:
        return v.upper() if v else v


class TariffUpdate(BaseModel):
    name:          Optional[str] = None
    description:   Optional[str] = None
    tariff_type:   Optional[str] = None
    supplier_id:   Optional[uuid.UUID] = None
    effective_from: Optional[date] = None
    effective_to:  Optional[date] = None
    unit:          Optional[str] = None
    currency_code: Optional[str] = None
    base_rate:     Optional[Decimal] = None
    peak_rate:     Optional[Decimal] = None
    offpeak_rate:  Optional[Decimal] = None
    demand_rate:   Optional[Decimal] = None
    fixed_charge:  Optional[Decimal] = None
    tax_rate_pct:  Optional[Decimal] = None
    tax_rule:      Optional[str] = None
    is_active:     Optional[bool] = None
    notes:         Optional[str] = None


class TariffRead(BaseModel):
    id:            uuid.UUID
    tariff_code:   str
    name:          str
    description:   Optional[str]
    utility_type:  str
    tariff_type:   str
    supplier_id:   Optional[uuid.UUID]
    effective_from: date
    effective_to:  Optional[date]
    unit:          str
    currency_code: str
    base_rate:     Decimal
    peak_rate:     Optional[Decimal]
    offpeak_rate:  Optional[Decimal]
    demand_rate:   Optional[Decimal]
    fixed_charge:  Optional[Decimal]
    tax_rate_pct:  Optional[Decimal]
    tax_rule:      Optional[str]
    is_active:     bool
    notes:         Optional[str]
    created_at:    Optional[str]
    updated_at:    Optional[str]

    model_config = {"from_attributes": True}


# ── Bill schemas ───────────────────────────────────────────────────────────────

class BillCreate(BaseModel):
    utility_type:        str
    provider_name:       Optional[str] = None
    supplier_id:         Optional[uuid.UUID] = None
    tariff_id:           Optional[uuid.UUID] = None
    tariff_code:         Optional[str] = None
    bill_reference:      Optional[str] = None
    billing_period_from: date
    billing_period_to:   date
    invoice_date:        Optional[date] = None
    due_date:            Optional[date] = None
    consumption_quantity: Optional[Decimal] = None
    consumption_unit:    Optional[str] = None
    unit_rate:           Optional[Decimal] = None
    base_charge:         Optional[Decimal] = None
    energy_charge:       Optional[Decimal] = None
    demand_charge:       Optional[Decimal] = None
    fixed_charge:        Optional[Decimal] = None
    surcharge_amount:    Optional[Decimal] = None
    tax_amount:          Optional[Decimal] = None
    total_amount:        Decimal
    currency_code:       str = "USD"
    status:              str = "RECEIVED"
    payment_date:        Optional[date] = None
    document_ref:        Optional[str] = None
    notes:               Optional[str] = None

    @field_validator("utility_type", "status", mode="before")
    @classmethod
    def upper(cls, v: str) -> str:
        return v.upper() if v else v


class BillUpdate(BaseModel):
    provider_name:       Optional[str] = None
    supplier_id:         Optional[uuid.UUID] = None
    tariff_id:           Optional[uuid.UUID] = None
    tariff_code:         Optional[str] = None
    bill_reference:      Optional[str] = None
    billing_period_from: Optional[date] = None
    billing_period_to:   Optional[date] = None
    invoice_date:        Optional[date] = None
    due_date:            Optional[date] = None
    consumption_quantity: Optional[Decimal] = None
    consumption_unit:    Optional[str] = None
    unit_rate:           Optional[Decimal] = None
    base_charge:         Optional[Decimal] = None
    energy_charge:       Optional[Decimal] = None
    demand_charge:       Optional[Decimal] = None
    fixed_charge:        Optional[Decimal] = None
    surcharge_amount:    Optional[Decimal] = None
    tax_amount:          Optional[Decimal] = None
    total_amount:        Optional[Decimal] = None
    currency_code:       Optional[str] = None
    status:              Optional[str] = None
    payment_date:        Optional[date] = None
    document_ref:        Optional[str] = None
    notes:               Optional[str] = None

    @field_validator("status", mode="before")
    @classmethod
    def upper(cls, v):
        return v.upper() if v else v


class BillRead(BaseModel):
    id:                  uuid.UUID
    bill_no:             str
    utility_type:        str
    provider_name:       Optional[str]
    supplier_id:         Optional[uuid.UUID]
    tariff_id:           Optional[uuid.UUID]
    tariff_code:         Optional[str]
    bill_reference:      Optional[str]
    billing_period_from: date
    billing_period_to:   date
    invoice_date:        Optional[date]
    due_date:            Optional[date]
    consumption_quantity: Optional[Decimal]
    consumption_unit:    Optional[str]
    unit_rate:           Optional[Decimal]
    base_charge:         Optional[Decimal]
    energy_charge:       Optional[Decimal]
    demand_charge:       Optional[Decimal]
    fixed_charge:        Optional[Decimal]
    surcharge_amount:    Optional[Decimal]
    tax_amount:          Optional[Decimal]
    total_amount:        Decimal
    currency_code:       str
    status:              str
    payment_date:        Optional[date]
    document_ref:        Optional[str]
    notes:               Optional[str]
    created_at:          Optional[str]
    updated_at:          Optional[str]

    model_config = {"from_attributes": True}


# ── Allocation schemas ─────────────────────────────────────────────────────────

class AllocationCreate(BaseModel):
    utility_type:        str
    allocation_date:     date
    period_start:        Optional[date] = None
    period_end:          Optional[date] = None
    allocation_method:   str = "METERED"
    target_type:         Optional[str] = None
    target_id:           Optional[str] = None
    target_name:         Optional[str] = None
    department:          Optional[str] = None
    production_line:     Optional[str] = None
    building_area:       Optional[str] = None
    machine_ref:         Optional[str] = None
    production_order_id: Optional[uuid.UUID] = None
    batch_no:            Optional[str] = None
    product_id:          Optional[uuid.UUID] = None
    source_cost:         Optional[Decimal] = None
    allocation_basis:    Optional[Decimal] = None
    allocation_basis_unit: Optional[str] = None
    allocated_quantity:  Optional[Decimal] = None
    quantity_unit:       Optional[str] = None
    production_volume_kg: Optional[Decimal] = None
    total_cost:          Decimal
    allocated_cost:      Decimal
    currency_code:       str = "USD"
    cost_per_unit:       Optional[Decimal] = None
    cost_per_ton:        Optional[Decimal] = None
    bill_id:             Optional[uuid.UUID] = None
    source_method:       str = "CALCULATED"
    notes:               Optional[str] = None

    @field_validator("utility_type", "allocation_method", "source_method", mode="before")
    @classmethod
    def upper(cls, v: str) -> str:
        return v.upper() if v else v

    @field_validator("target_type", mode="before")
    @classmethod
    def upper_opt(cls, v):
        return v.upper() if v else v

    @model_validator(mode="after")
    def compute_derived(self) -> "AllocationCreate":
        # Auto-compute cost_per_unit if missing
        if self.cost_per_unit is None and self.allocated_quantity and self.allocated_quantity > 0:
            self.cost_per_unit = self.allocated_cost / self.allocated_quantity
        # Auto-compute cost_per_ton if missing
        if self.cost_per_ton is None and self.production_volume_kg and self.production_volume_kg > 0:
            self.cost_per_ton = self.allocated_cost / (self.production_volume_kg / 1000)
        return self


class AllocationUpdate(BaseModel):
    allocation_method:   Optional[str] = None
    target_type:         Optional[str] = None
    target_id:           Optional[str] = None
    target_name:         Optional[str] = None
    department:          Optional[str] = None
    production_line:     Optional[str] = None
    building_area:       Optional[str] = None
    machine_ref:         Optional[str] = None
    production_order_id: Optional[uuid.UUID] = None
    batch_no:            Optional[str] = None
    product_id:          Optional[uuid.UUID] = None
    source_cost:         Optional[Decimal] = None
    allocation_basis:    Optional[Decimal] = None
    allocation_basis_unit: Optional[str] = None
    allocated_quantity:  Optional[Decimal] = None
    quantity_unit:       Optional[str] = None
    production_volume_kg: Optional[Decimal] = None
    total_cost:          Optional[Decimal] = None
    allocated_cost:      Optional[Decimal] = None
    currency_code:       Optional[str] = None
    cost_per_unit:       Optional[Decimal] = None
    cost_per_ton:        Optional[Decimal] = None
    bill_id:             Optional[uuid.UUID] = None
    is_approved:         Optional[bool] = None
    notes:               Optional[str] = None


class AllocationRead(BaseModel):
    id:                  uuid.UUID
    allocation_no:       str
    utility_type:        str
    allocation_date:     date
    period_start:        Optional[date]
    period_end:          Optional[date]
    allocation_method:   str
    target_type:         Optional[str]
    target_id:           Optional[str]
    target_name:         Optional[str]
    department:          Optional[str]
    production_line:     Optional[str]
    building_area:       Optional[str]
    machine_ref:         Optional[str]
    production_order_id: Optional[uuid.UUID]
    batch_no:            Optional[str]
    product_id:          Optional[uuid.UUID]
    source_cost:         Optional[Decimal]
    allocation_basis:    Optional[Decimal]
    allocation_basis_unit: Optional[str]
    allocated_quantity:  Optional[Decimal]
    quantity_unit:       Optional[str]
    production_volume_kg: Optional[Decimal]
    total_cost:          Decimal
    allocated_cost:      Decimal
    currency_code:       str
    cost_per_unit:       Optional[Decimal]
    cost_per_ton:        Optional[Decimal]
    bill_id:             Optional[uuid.UUID]
    source_method:       str
    is_approved:         bool
    notes:               Optional[str]
    created_at:          Optional[str]
    updated_at:          Optional[str]

    model_config = {"from_attributes": True}


# ── KPIs & reports ─────────────────────────────────────────────────────────────

class BillingKPIs(BaseModel):
    date_from:             Optional[str] = None
    date_to:               Optional[str] = None
    utility_type:          Optional[str] = None
    total_bills:           int
    total_billed_amount:   Decimal
    total_tax_amount:      Optional[Decimal]
    total_consumption:     Optional[Decimal]
    consumption_unit:      Optional[str]
    avg_unit_rate:         Optional[Decimal]
    unpaid_amount:         Optional[Decimal]
    overdue_count:         int
    disputed_count:        int
    currency_code:         str = "USD"


class AllocationReportRow(BaseModel):
    target_type:      Optional[str]
    target_id:        Optional[str]
    target_name:      Optional[str]
    utility_type:     Optional[str]
    allocated_cost:   Decimal
    allocated_quantity: Optional[Decimal]
    quantity_unit:    Optional[str]
    cost_per_unit:    Optional[Decimal]
    record_count:     int


class AllocationReport(BaseModel):
    period_start:    Optional[str]
    period_end:      Optional[str]
    utility_type:    Optional[str]
    group_by:        str
    rows:            List[AllocationReportRow]
    grand_total:     Decimal


# ── Allocation engine request ──────────────────────────────────────────────────

class AllocationEngineTarget(BaseModel):
    target_type:     str
    target_id:       str
    target_name:     Optional[str] = None
    basis_value:     Decimal   # e.g. runtime hours, %, m², etc.
    basis_unit:      Optional[str] = None


class AllocationEngineRequest(BaseModel):
    bill_id:           uuid.UUID
    allocation_method: str
    allocation_date:   date
    currency_code:     str = "USD"
    targets:           List[AllocationEngineTarget]

    @field_validator("allocation_method", mode="before")
    @classmethod
    def upper(cls, v: str) -> str:
        return v.upper() if v else v


class AllocationEngineResult(BaseModel):
    bill_id:           uuid.UUID
    bill_no:           str
    total_cost:        Decimal
    allocation_method: str
    allocations_created: int
    rows:              List[AllocationRead]

File: frontend/src/lib/utilityBilling.ts
/**
 * Utility Billing, Tariffs & Cost Allocation — TypeScript types and API client.
 */
import { apiClient } from "./api";

// ── Enums & label maps ────────────────────────────────────────────────────────

export type UtilityTypeBilling =
  | "ELECTRICITY" | "WATER" | "SOFT_WATER" | "PROCESS_WATER"
  | "STEAM" | "COMPRESSED_AIR" | "NATURAL_GAS" | "SOLAR"
  | "WASTEWATER" | "DIESEL" | "CHILLED_WATER" | "OTHER";

export const UTILITY_TYPE_LABELS: Record<UtilityTypeBilling, string> = {
  ELECTRICITY:    "Electricity",
  WATER:          "Water",
  SOFT_WATER:     "Soft Water",
  PROCESS_WATER:  "Process Water",
  STEAM:          "Steam",
  COMPRESSED_AIR: "Compressed Air",
  NATURAL_GAS:    "Natural Gas",
  SOLAR:          "Solar",
  WASTEWATER:     "Wastewater",
  DIESEL:         "Diesel",
  CHILLED_WATER:  "Chilled Water",
  OTHER:          "Other",
};

export const UTILITY_TYPE_COLORS: Record<UtilityTypeBilling, string> = {
  ELECTRICITY:    "#fbbf24",
  WATER:          "#60a5fa",
  SOFT_WATER:     "#93c5fd",
  PROCESS_WATER:  "#3b82f6",
  STEAM:          "#f97316",
  COMPRESSED_AIR: "#22d3ee",
  NATURAL_GAS:    "#a78bfa",
  SOLAR:          "#facc15",
  WASTEWATER:     "#6b7280",
  DIESEL:         "#94a3b8",
  CHILLED_WATER:  "#67e8f9",
  OTHER:          "#9ca3af",
};

export type BillStatus = "DRAFT" | "RECEIVED" | "VERIFIED" | "PAID" | "DISPUTED" | "OVERDUE" | "VOID";

export const BILL_STATUS_LABELS: Record<BillStatus, string> = {
  DRAFT:    "Draft",
  RECEIVED: "Received",
  VERIFIED: "Verified",
  PAID:     "Paid",
  DISPUTED: "Disputed",
  OVERDUE:  "Overdue",
  VOID:     "Void",
};

export const BILL_STATUS_COLORS: Record<BillStatus, string> = {
  DRAFT:    "#6b7280",
  RECEIVED: "#3b82f6",
  VERIFIED: "#8b5cf6",
  PAID:     "#22c55e",
  DISPUTED: "#f97316",
  OVERDUE:  "#ef4444",
  VOID:     "#9ca3af",
};

export type TariffType = "FLAT" | "TIME_OF_USE" | "DEMAND" | "STEPPED";

export const TARIFF_TYPE_LABELS: Record<TariffType, string> = {
  FLAT:        "Flat Rate",
  TIME_OF_USE: "Time-of-Use",
  DEMAND:      "Demand",
  STEPPED:     "Stepped / Block",
};

export type AllocationMethod =
  | "METERED" | "PROPORTIONAL" | "FIXED" | "CALCULATED"
  | "RUNTIME" | "STANDARD_CONSUMPTION" | "COST_CENTER";

export const ALLOCATION_METHOD_LABELS: Record<AllocationMethod, string> = {
  METERED:              "Direct Metered",
  PROPORTIONAL:         "Proportional",
  FIXED:                "Fixed",
  CALCULATED:           "Calculated",
  RUNTIME:              "Est. by Runtime",
  STANDARD_CONSUMPTION: "Est. by Std. Consumption",
  COST_CENTER:          "Cost-Center Based",
};

export type TargetType =
  | "DEPARTMENT" | "PRODUCTION_LINE" | "MACHINE" | "PRODUCT"
  | "PRODUCTION_ORDER" | "BATCH" | "SUPPORT_AREA" | "OVERHEAD_POOL";

export const TARGET_TYPE_LABELS: Record<TargetType, string> = {
  DEPARTMENT:       "Department",
  PRODUCTION_LINE:  "Production Line",
  MACHINE:          "Machine",
  PRODUCT:          "Product",
  PRODUCTION_ORDER: "Production Order",
  BATCH:            "Batch",
  SUPPORT_AREA:     "Support Area",
  OVERHEAD_POOL:    "Overhead Pool",
};

// ── Tariff types ──────────────────────────────────────────────────────────────

export interface UtilityTariff {
  id:            string;
  tariff_code:   string;
  name:          string;
  description?:  string | null;
  utility_type:  UtilityTypeBilling;
  tariff_type:   TariffType;
  supplier_id?:  string | null;
  effective_from: string;
  effective_to?:  string | null;
  unit:          string;
  currency_code: string;
  base_rate:     number;
  peak_rate?:    number | null;
  offpeak_rate?: number | null;
  demand_rate?:  number | null;
  fixed_charge?: number | null;
  tax_rate_pct?: number | null;
  tax_rule?:     string | null;
  is_active:     boolean;
  notes?:        string | null;
  created_at?:   string | null;
  updated_at?:   string | null;
}

export type TariffCreate = Omit<UtilityTariff, "id" | "created_at" | "updated_at">;
export type TariffUpdate = Partial<Omit<UtilityTariff, "id" | "tariff_code" | "utility_type" | "created_at" | "updated_at">>;

// ── Bill types ────────────────────────────────────────────────────────────────

export interface UtilityBill {
  id:                  string;
  bill_no:             string;
  utility_type:        UtilityTypeBilling;
  provider_name?:      string | null;
  supplier_id?:        string | null;
  tariff_id?:          string | null;
  tariff_code?:        string | null;
  bill_reference?:     string | null;
  billing_period_from: string;
  billing_period_to:   string;
  invoice_date?:       string | null;
  due_date?:           string | null;
  consumption_quantity?: number | null;
  consumption_unit?:   string | null;
  unit_rate?:          number | null;
  base_charge?:        number | null;
  energy_charge?:      number | null;
  demand_charge?:      number | null;
  fixed_charge?:       number | null;
  surcharge_amount?:   number | null;
  tax_amount?:         number | null;
  total_amount:        number;
  currency_code:       string;
  status:              BillStatus;
  payment_date?:       string | null;
  document_ref?:       string | null;
  notes?:              string | null;
  created_at?:         string | null;
  updated_at?:         string | null;
}

export type BillCreate = Omit<UtilityBill, "id" | "bill_no" | "created_at" | "updated_at">;
export type BillUpdate = Partial<Omit<UtilityBill, "id" | "bill_no" | "utility_type" | "created_at" | "updated_at">>;

// ── Allocation types ──────────────────────────────────────────────────────────

export interface UtilityAllocation {
  id:                    string;
  allocation_no:         string;
  utility_type:          UtilityTypeBilling;
  allocation_date:       string;
  period_start?:         string | null;
  period_end?:           string | null;
  allocation_method:     AllocationMethod;
  target_type?:          TargetType | null;
  target_id?:            string | null;
  target_name?:          string | null;
  department?:           string | null;
  production_line?:      string | null;
  building_area?:        string | null;
  machine_ref?:          string | null;
  production_order_id?:  string | null;
  batch_no?:             string | null;
  product_id?:           string | null;
  source_cost?:          number | null;
  allocation_basis?:     number | null;
  allocation_basis_unit?: string | null;
  allocated_quantity?:   number | null;
  quantity_unit?:        string | null;
  production_volume_kg?: number | null;
  total_cost:            number;
  allocated_cost:        number;
  currency_code:         string;
  cost_per_unit?:        number | null;
  cost_per_ton?:         number | null;
  bill_id?:              string | null;
  source_method:         string;
  is_approved:           boolean;
  notes?:                string | null;
  created_at?:           string | null;
  updated_at?:           string | null;
}

export type AllocationCreate = Omit<UtilityAllocation, "id" | "allocation_no" | "is_approved" | "created_at" | "updated_at">;
export type AllocationUpdate = Partial<Omit<UtilityAllocation, "id" | "allocation_no" | "utility_type" | "allocation_date" | "created_at" | "updated_at">>;

// ── KPI & report types ────────────────────────────────────────────────────────

export interface BillingKPIs {
  date_from?:           string | null;
  date_to?:             string | null;
  utility_type?:        string | null;
  total_bills:          number;
  total_billed_amount:  number;
  total_tax_amount?:    number | null;
  total_consumption?:   number | null;
  consumption_unit?:    string | null;
  avg_unit_rate?:       number | null;
  unpaid_amount?:       number | null;
  overdue_count:        number;
  disputed_count:       number;
  currency_code:        string;
}

export interface AllocationReportRow {
  target_type?:      string | null;
  target_id?:        string | null;
  target_name?:      string | null;
  utility_type?:     string | null;
  allocated_cost:    number;
  allocated_quantity?: number | null;
  quantity_unit?:    string | null;
  cost_per_unit?:    number | null;
  record_count:      number;
}

export interface AllocationReport {
  period_start?:  string | null;
  period_end?:    string | null;
  utility_type?:  string | null;
  group_by:       string;
  rows:           AllocationReportRow[];
  grand_total:    number;
}

// ── Engine types ──────────────────────────────────────────────────────────────

export interface AllocationEngineTarget {
  target_type:   TargetType;
  target_id:     string;
  target_name?:  string | null;
  basis_value:   number;
  basis_unit?:   string | null;
}

export interface AllocationEngineRequest {
  bill_id:           string;
  allocation_method: AllocationMethod;
  allocation_date:   string;
  currency_code?:    string;
  targets:           AllocationEngineTarget[];
}

export interface AllocationEngineResult {
  bill_id:             string;
  bill_no:             string;
  total_cost:          number;
  allocation_method:   string;
  allocations_created: number;
  rows:                UtilityAllocation[];
}

// ── Filter types ──────────────────────────────────────────────────────────────

export interface BillFilters {
  utility_type?:  UtilityTypeBilling;
  status?:        BillStatus;
  date_from?:     string;
  date_to?:       string;
  supplier_id?:   string;
  skip?:          number;
  limit?:         number;
}

export interface TariffFilters {
  utility_type?:  UtilityTypeBilling;
  is_active?:     boolean;
  as_of_date?:    string;
  skip?:          number;
  limit?:         number;
}

export interface AllocationFilters {
  utility_type?:    UtilityTypeBilling;
  target_type?:     TargetType;
  target_id?:       string;
  department?:      string;
  production_line?: string;
  machine_ref?:     string;
  batch_no?:        string;
  bill_id?:         string;
  date_from?:       string;
  date_to?:         string;
  is_approved?:     boolean;
  skip?:            number;
  limit?:           number;
}

// ── CSV download ──────────────────────────────────────────────────────────────

export async function downloadBillsCsv(
  params: Record<string, string | number | boolean | undefined>,
  filename = "utility_bills.csv",
): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const qs    = new URLSearchParams(
    Object.entries(params).filter(([, v]) => v !== undefined && v !== "").map(([k, v]) => [k, String(v)])
  ).toString();
  const res = await fetch(`${base}/api/v1/billing/bills/export/csv${qs ? "?" + qs : ""}`,
    { credentials: "include" });
  if (!res.ok) throw new Error(`Export failed (${res.status})`);
  const blob = await res.blob();
  const url  = URL.createObjectURL(blob);
  const a = document.createElement("a"); a.href = url; a.download = filename;
  document.body.appendChild(a); a.click(); document.body.removeChild(a); URL.revokeObjectURL(url);
}

export async function downloadAllocationsCsv(
  params: Record<string, string | number | boolean | undefined>,
  filename = "cost_allocations.csv",
): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const qs    = new URLSearchParams(
    Object.entries(params).filter(([, v]) => v !== undefined && v !== "").map(([k, v]) => [k, String(v)])
  ).toString();
  const res = await fetch(`${base}/api/v1/billing/allocations/export/csv${qs ? "?" + qs : ""}`,
    { credentials: "include" });
  if (!res.ok) throw new Error(`Export failed (${res.status})`);
  const blob = await res.blob();
  const url  = URL.createObjectURL(blob);
  const a = document.createElement("a"); a.href = url; a.download = filename;
  document.body.appendChild(a); a.click(); document.body.removeChild(a); URL.revokeObjectURL(url);
}

// ── API client ────────────────────────────────────────────────────────────────

export const billingApi = {
  // Tariffs
  async listTariffs(params?: TariffFilters): Promise<UtilityTariff[]> {
    const r = await apiClient.get<UtilityTariff[]>("/api/v1/billing/tariffs", { params });
    return r.data;
  },
  async createTariff(data: TariffCreate): Promise<UtilityTariff> {
    const r = await apiClient.post<UtilityTariff>("/api/v1/billing/tariffs", data);
    return r.data;
  },
  async updateTariff(id: string, data: TariffUpdate): Promise<UtilityTariff> {
    const r = await apiClient.patch<UtilityTariff>(`/api/v1/billing/tariffs/${id}`, data);
    return r.data;
  },
  async deleteTariff(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/billing/tariffs/${id}`).then((r) => r.data);
  },
  async lookupTariff(utilityType: string, asOf?: string): Promise<UtilityTariff | null> {
    const r = await apiClient.get<UtilityTariff | null>("/api/v1/billing/tariffs/lookup",
      { params: { utility_type: utilityType, as_of: asOf } });
    return r.data;
  },

  // Bills
  async kpis(params?: Partial<BillFilters & { currency_code?: string }>): Promise<BillingKPIs> {
    const r = await apiClient.get<BillingKPIs>("/api/v1/billing/bills/kpis", { params });
    return r.data;
  },
  async listBills(params?: BillFilters): Promise<UtilityBill[]> {
    const r = await apiClient.get<UtilityBill[]>("/api/v1/billing/bills", { params });
    return r.data;
  },
  async createBill(data: BillCreate): Promise<UtilityBill> {
    const r = await apiClient.post<UtilityBill>("/api/v1/billing/bills", data);
    return r.data;
  },
  async updateBill(id: string, data: BillUpdate): Promise<UtilityBill> {
    const r = await apiClient.patch<UtilityBill>(`/api/v1/billing/bills/${id}`, data);
    return r.data;
  },
  async deleteBill(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/billing/bills/${id}`).then((r) => r.data);
  },

  // Allocations
  async listAllocations(params?: AllocationFilters): Promise<UtilityAllocation[]> {
    const r = await apiClient.get<UtilityAllocation[]>("/api/v1/billing/allocations", { params });
    return r.data;
  },
  async createAllocation(data: AllocationCreate): Promise<UtilityAllocation> {
    const r = await apiClient.post<UtilityAllocation>("/api/v1/billing/allocations", data);
    return r.data;
  },
  async updateAllocation(id: string, data: AllocationUpdate): Promise<UtilityAllocation> {
    const r = await apiClient.patch<UtilityAllocation>(`/api/v1/billing/allocations/${id}`, data);
    return r.data;
  },
  async deleteAllocation(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/billing/allocations/${id}`).then((r) => r.data);
  },
  async runEngine(req: AllocationEngineRequest): Promise<AllocationEngineResult> {
    const r = await apiClient.post<AllocationEngineResult>("/api/v1/billing/allocations/run-engine", req);
    return r.data;
  },
  async report(params?: {
    group_by?: string; utility_type?: string; date_from?: string; date_to?: string;
  }): Promise<AllocationReport> {
    const r = await apiClient.get<AllocationReport>("/api/v1/billing/allocations/report", { params });
    return r.data;
  },
};
